### DISCUSSION TITLE
Adopt the Robustness Principle going forward

### DISCUSSION THREAD

**Discussion by dogweather** [sqlalchemy/sqlalchemy#8437] [Usage Questions]
### Describe the bug

Apps around the world are breaking because auto-generated URLs with `postgres://` are no longer accepted. I submit, it doesn't have to be this way:

* There isn't a `postgres` DBMS that must to be differentiated from `postgresql`.
* "Postgres" is a common (even if incorrect) name for the DBMS.
* A major, highly respected principle in engineering is, [be conservative in what you send, be liberal in what you accept](https://en.wikipedia.org/wiki/Robustness_principle).
* It's not SQL Alchemy's job to teach people — painfully — PostgreSQL's proper name.
* Our overly strict API has resulted in conflicting blog and SO posts attempting to handle the problem, such as [this one](https://stackoverflow.com/questions/62688256/sqlalchemy-exc-nosuchmoduleerror-cant-load-plugin-sqlalchemy-dialectspostgre). IMO this is all wasted effort and time.

### Proposed solution

At this public interface, we should be liberal in what we accept, and accept `postgres`. There are no downsides and only upsides.

### To Reproduce

Create an app on nearly any service or tool, such as Heroku or Render.com. It will fail because these auto-generate `postgres://` URLs.

### Error

```
sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:postgres
``` 

### Versions

- OS: Ubuntu
- Python: 3.10.5
- SQLAlchemy: 1.4.40
- Database: PostgreSQL 13 on Heroku and Render.com.
- DBAPI: `postgresql`

**Comment by CaselIT** [comment]
Hi,

as mentioned in multiple other settings, like https://github.com/sqlalchemy/sqlalchemy/issues/6083#issuecomment-801478013 and https://github.com/pallets-eco/flask-sqlalchemy/issues/929#issuecomment-801049996 the `postgres` name has been deprecated since 2010 years and is was raising a warning since at least 2016.

why no app around the world ever opened an issue saying "hey my app still needs `postgres` since it's used by my favorite hosting provider"?

we honestly did not think it was a big deal when it was removed, since no one ever complained that it was being deprecated

I'm not familiar with the linked rodustness principle but if it imply that nothing can be deprecated and then removed, ever, then I don't thing it would be something we are interested in adopting, sorry

**Reply by jvanasco** [reply]
Beyond what @CaselIT noted and the discussion in the other threads, I'd like to remind everyone the deprecation was fully in line with the tenets of [PEP 20](https://peps.python.org/pep-0020/)

**Comment by zzzeek** [comment]
> 
>     * A major, highly respected principle in engineering is, [be conservative in what you send, be liberal in what you accept](https://en.wikipedia.org/wiki/Robustness_principle).

Well if people no longer respect SQLAlchemy because we dropped a name that was explicitly deprecated for 11 years, so be it.  The architecture of our dialect lookup was highly interrupted by having this one odd name stuck in there, and there's just no reason for us to have this one special name sticking out there for over 11 years.     We deprecate old names and methods all the time, and old names and methods get removed all the time, it's just in this case, some documentation like that of Heroku seemed to have the old name hardcoded there, so it caused a lot of problems.

>     * It's not SQL Alchemy's job to teach people — painfully — PostgreSQL's proper name.

It's people who use SQLAlchemy's job to heed our deprecation warnings, this one has been emitting a warning since (checks) 2016.  that's six years ago.

> 
>     * Our overly strict API has resulted in conflicting blog and SO posts attempting to handle the problem, such as [this one](https://stackoverflow.com/questions/62688256/sqlalchemy-exc-nosuchmoduleerror-cant-load-plugin-sqlalchemy-dialectspostgre). 

I've read every answer on that SO question and every single answer seems to correctly indicate, "change postgres to postgresql", where is the conflict?

> IMO this is all wasted effort and time.

having to work around one odd name stuck in the middle of the code, which meant it had to have it's own hardcoded rule and all that, was wasting our effort and time, so we removed it, after telling people for eleven years and warning at runtime for six years that we were going to do it.  We thought that was enough time.   But that horse has left the barn as of 1.4's first release almost two years ago, the vast majority of people affected by this have migrated, I just read today that Heroku is deleting all the "free" accounts anyway, and my experience in these matters is that even a deprecation that's a little bumpy, which this one is due to the popularity of Heroku and their incorrect documentation, eventually everyone gets migrated and the problem is solved.   I would assume you read all the discussion here, got a feel for where we are on this, and you are...expecting a sudden change here on this end?    I've had to deal with much more nasty API changes than this in my career and certainly, they don't get reversed...this is regular software development stuff.  name was warning deprecated for six years.  that's generous.   

> ### To Reproduce
> 
> Create an app on nearly any service or tool, such as Heroku or Render.com. It will fail because these auto-generate `postgres://` URLs.

That is a bug in Heroku.   They told us over and over again that this is a documentation issue, and nothing on their end they can change.    You are saying here that creating an app somehow generates a broken SQLAlchemy URL.   I'm sort of mystified if that's true but, complain to them. You're a paying customer now, they really should fix this.  No idea how they could have a broken URL in their docs/code/plugins/recipes/whatever at this point.      I'm sure "we fix bugs in our code" is also a programming principle we can agree on.....Heroku should get on that.

**Reply by dogweather** [reply]
My focus is on the Robustness Principle in the public API design. I guess you're not in favor of that?

**Reply by zzzeek** [reply]
The Robustness Principle: "be conservative in what you do, be liberal in what you accept from others".   "Conservative" and "Liberal" are subjective terms and can be made to mean anything.  We certainly follow the former.  The "be liberal in what you accept", that can mean a lot of things but in the Python world, we actually have somewhat of a culture in not being *too* liberal, because accepting all APIs in perpetuity lest someone is inconvenienced is widely accepted as a terrible idea in software devvelopment.   Python deprecates and outright removes [major, foundational public APIs that the entire ecosystem is built upon](https://peps.python.org/pep-0632/) sometimes.  This is normal.    

I really think you need to consider **this is a heroku problem**.  They have an incorrect keyword in their code, **for eleven years**.   nobody else has ever complained about this issue in any way.    The SQLAlchemy project is quite simply not in the business of servicing giant corporations who refuse to make the slightest effort to solve their own problems.

**Comment by jvanasco** [comment]
> It's people who use SQLAlchemy's job to heed our deprecation warnings, this one has been emitting a warning since (checks) 2016. that's six years ago.

Incorrect, Mike.  Warnings have been emitted since 2009. That is 13 years ago (https://github.com/sqlalchemy/sqlalchemy/blob/8fc5005dfe3eb66a46470ad8a8c7b95fc4d6bdca/lib/sqlalchemy/dialects/postgres.py)

In 2016 there was an internal architecture change: the deprecated file was removed and the temp support was handled by that translation tool you made (https://github.com/sqlalchemy/sqlalchemy/commit/c59bf0007f0148b73c80a3d6b86d6c66ae0ab422#diff-a453b1c3c392ab7bd09e97ab080042e9e4d1485f107b99011a39729c561004ba)
